relay-replays/src/recording/protocol.rs

```diff
@@ -0,0 +1,171 @@
+/// Protocol.
+///
+/// The following section includes the behaviors responsible for deserializing a recording from
+/// its HTTP transport schema and serializing a recording to its Kafka transport schema.
+///
+/// We expect recordings to come in the format of plaintext headers (JSON encoded), then a new
+/// line character, then an optionally compressed RRWeb recording.  Failure to include any
+/// component of this schema results in an error.
+use flate2::{read::ZlibDecoder, write::ZlibEncoder, Compression};
+use serde::Deserialize;
+use serde_json::Error as SError;
+use std::fmt::Display;
+use std::io::{Error, Read, Write};
+
+/// Deserializes input bytes and returns a tuple of serialized (confirmed valid) JSON headers
+/// and the optionally decompressed body value.
+pub fn deserialize(bytes: &[u8], limit: usize) -> Result<(&[u8], Vec<u8>), ProtocolError> {
+    let (headers, body) = read(bytes)?;
+
+    // We always attempt to decompress the body value. If decompression fails we try to JSON
+    // deserialize the body bytes as is.
+    match decompress(body, limit) {
+        Ok(buf) => Ok((headers, buf)),
+        Err(_) => Ok((headers, body.into())),
+    }
+}
+
+/// Serializes the headers and body arguments into a single vec of bytes. The body value is
+/// compressed before being concatenated.  The final output is headers-bytes + "\n" (new-line
+/// character) + compressed-body-bytes.
+pub fn serialize(headers: &[u8], body: Vec<u8>) -> Result<Vec<u8>, ProtocolError> {
+    let output_bytes = compress(body)?;
+    Ok([headers.into(), vec![b'\n'], output_bytes].concat())
+}
+
+fn compress(output: Vec<u8>) -> Result<Vec<u8>, ProtocolError> {
+    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
+    encoder
+        .write_all(&output)
+        .map_err(ProtocolError::EncodingError)?;
+    encoder.finish().map_err(ProtocolError::EncodingError)
+}
+
+fn decompress(zipped_bytes: &[u8], limit: usize) -> Result<Vec<u8>, ProtocolError> {
+    let mut buffer = Vec::new();
+
+    let decoder = ZlibDecoder::new(zipped_bytes);
+    decoder
+        .take(limit as u64)
+        .read_to_end(&mut buffer)
+        .map_err(ProtocolError::EncodingError)?;
+
+    Ok(buffer)
+}
+
+fn read(bytes: &[u8]) -> Result<(&[u8], &[u8]), ProtocolError> {
+    match bytes.is_empty() {
+        true => Err(ProtocolError::MissingData),
+        false => {
```

Comment:
Please do not match on a bool, use `if` instead.

```suggestion
    if bytes.is_empty() {
        return Err(ProtocolError::MissingData);
    }
    
    // ...
```